core/primitives/src/epoch_manager.rs

```diff
@@ -49,117 +40,975 @@ pub struct EpochConfig {
+#[cfg(feature = "protocol_feature_block_header_v3")]
+pub mod epoch_info {
+    use crate::epoch_manager::ValidatorWeight;
+    use crate::types::validator_stake::{ValidatorStake, ValidatorStakeIter, ValidatorStakeV1};
+    use crate::types::{BlockChunkValidatorStats, ValidatorKickoutReason};
+    use crate::version::PROTOCOL_VERSION;
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use near_primitives_core::hash::CryptoHash;
+    use near_primitives_core::types::{
+        AccountId, Balance, EpochHeight, ProtocolVersion, ValidatorId,
+    };
+    use serde::Serialize;
+    use smart_default::SmartDefault;
+    use std::collections::{BTreeMap, HashMap};
+
+    /// Information per epoch.
+    #[derive(BorshSerialize, BorshDeserialize, Serialize, Clone, Debug, PartialEq, Eq)]
+    pub enum EpochInfo {
+        V1(EpochInfoV1),
+        V2(EpochInfoV2),
+    }
+
+    impl Default for EpochInfo {
+        fn default() -> Self {
+            Self::V2(EpochInfoV2::default())
+        }
+    }
+
+    #[derive(
+        SmartDefault, BorshSerialize, BorshDeserialize, Serialize, Clone, Debug, PartialEq, Eq,
+    )]
+    pub struct EpochInfoV1 {
+        /// Ordinal of given epoch from genesis.
+        /// There can be multiple epochs with the same ordinal in case of long forks.
+        pub epoch_height: EpochHeight,
+        /// List of current validators.
+        pub validators: Vec<ValidatorStakeV1>,
+        /// Validator account id to index in proposals.
+        pub validator_to_index: HashMap<AccountId, ValidatorId>,
+        /// Settlement of validators responsible for block production.
+        pub block_producers_settlement: Vec<ValidatorId>,
+        /// Per each shard, settlement validators that are responsible.
+        pub chunk_producers_settlement: Vec<Vec<ValidatorId>>,
+        /// Settlement of hidden validators with weights used to determine how many shards they will validate.
+        pub hidden_validators_settlement: Vec<ValidatorWeight>,
+        /// List of current fishermen.
+        pub fishermen: Vec<ValidatorStakeV1>,
+        /// Fisherman account id to index of proposal.
+        pub fishermen_to_index: HashMap<AccountId, ValidatorId>,
+        /// New stake for validators.
+        pub stake_change: BTreeMap<AccountId, Balance>,
+        /// Validator reward for the epoch.
+        pub validator_reward: HashMap<AccountId, Balance>,
+        /// Validators who are kicked out in this epoch.
+        pub validator_kickout: HashMap<AccountId, ValidatorKickoutReason>,
+        /// Total minted tokens in the epoch.
+        pub minted_amount: Balance,
+        /// Seat price of this epoch.
+        pub seat_price: Balance,
+        /// Current protocol version during this epoch.
+        #[default(PROTOCOL_VERSION)]
+        pub protocol_version: ProtocolVersion,
+    }
+
+    #[derive(
+        SmartDefault, BorshSerialize, BorshDeserialize, Serialize, Clone, Debug, PartialEq, Eq,
+    )]
+    pub struct EpochInfoV2 {
```

Comment:
Same here

Reply:
Done in https://github.com/near/nearcore/pull/4115/commits/a9f5f9004348c8e8175d8e974303f682d06b3942